### PR TITLE
Fixes Process listeners after refresh()

### LINKED ISSUE #1401: Realm.addChangeListener doesn't work on background threads
If you do 

```
new Thread(new Runnable() {
  public void run() {
    Realm realm = Realm.getDefaultInstance();
    realm.addChangeListener(listener)
    // ...
  }

}).start();
```

on a non-looper thread. That change listener will never be called unless you execute a transaction on that thread.

That behaviour seems a bit strange. The problem is that we are triggering onChange after updating the the Realm on looper-threads. That isn't possible on non-looper threads as there is no point in time updating the Realm behind users back would make sense.

We should consider doing one of the following:

1) Prevent people from adding change listeners on a non-looper thread (throw exception). People would have to notify these threads in some other way.

2) Document the behaviour better. 

3) Call onChange on non-looper threads, but without updating the Realm on that thread first.

4) Something else.... ?

### PR DESCRIPTION
**Pull request by emanuelez**
Fixes #1401

### CODE REVIEW

**Comment by emanuelez** [comment]
@realm/java

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 1]
```diff
@@ -1818,4 +1822,40 @@ public void testDeletingRealmAlsoClearsConfigurationCache() throws IOException {
         testRealm = Realm.getInstance(newConfig);
         assertNotNull(testRealm);
     }
+
+    public void testProcessLocalListenersAfterRefresh() {
+        // Used to validate the result
+        final AtomicBoolean listenerWasCalled = new AtomicBoolean(false);
+
+        // Used by the background thread to wait for the main thread to do the write operation
+        final CountDownLatch latch = new CountDownLatch(1);
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Perhaps call this `bgThreadReady` or something. `latch` is a bit too generic imo.

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 2]
```diff
@@ -1818,4 +1822,40 @@ public void testDeletingRealmAlsoClearsConfigurationCache() throws IOException {
...
+                    bgRealm.close();
+                }
+            }
+        };
+        backgroundThread.start();
+
+        testRealm.beginTransaction();
+        testRealm.createObject(Dog.class);
+        testRealm.commitTransaction();
+        latch.countDown();
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Does this actually work? Doesn't the test just complete now, then ignoring any failures from running background threads?

File: realm/src/main/java/io/realm/Realm.java [hunk 3]
```diff
@@ -1276,6 +1276,7 @@ boolean hasChanged() {
     public void refresh() {
         checkIfValid();
         transaction.advanceRead();
+        sendNotifications();
```
Comment (beeender) [inline realm/src/main/java/io/realm/Realm.java]: Then every thread will be notified? Which means the thread with looper will be notified more than once with nothing changed.

**Comment by cmelchior** [comment]
Also. It would probably be nice to expand the JavaDoc to document the slightly different behaviour for listeners in `addChangeListener` and `refresh`

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 2 above)
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: It does work, in fact the test fails without the change in `Realm.java`

File: realm/src/main/java/io/realm/Realm.java (hunk 3 above)
Comment (emanuelez) [inline realm/src/main/java/io/realm/Realm.java]: no, sendNotifications operates on the thread local list of changeListeners

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 1 above)
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: it's a latch, not a boolean. How about `bgThreadLatch`?

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 2 above)
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: :+1:

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 1 above)
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: :+1:

File: realm/src/main/java/io/realm/Realm.java (hunk 3 above)
Comment (beeender) [inline realm/src/main/java/io/realm/Realm.java]: I see...
Is it possible to check the version then call `sendNotification`? I have seen @cmelchior made a change to return the version by `advanceRead()`

Comment (emanuelez) [inline realm/src/main/java/io/realm/Realm.java]: Right now `advanceRead()` returns `void`, but that would be a nice addition once available!

Comment (cmelchior) [inline realm/src/main/java/io/realm/Realm.java]: As requested :) https://github.com/realm/realm-java/pull/1408

**Comment by beeender** [comment]
:+1: 
Just a reminder, need one more test case (refresh doesn't trigger when nothing changed.) when #1408 merged.

**Comment by emanuelez** [comment]
retest this please

**Comment by emanuelez** [comment]
Please ignore the failing dco-signed step... it's my experiments with Leeroy :)

**Comment by emanuelez** [comment]
Changelog:
 * Fixed an issue preventing the call of listeners on refresh()

**Comment by emanuelez** [comment]
Bump @realm/java

**Comment by beeender** [comment]
:+1: again

**Comment by kneth** [comment]
:+1: But I would like to see #1408 merged first.

**Comment by cmelchior** [comment]
#1408 is not correct. After some digging, i discovered that the reason my test case fail is because I just use an anonymous change listener instance, which then gets GC'ed at a random point in time because we use WeakReferences....I hate WeakReferences :(

:+1:

**Comment by emanuelez** [comment]
I assume this is ok to merge then, right @cmelchior?

**Comment by cmelchior** [comment]
Yup :+1:

**Pull request merged**